Add accommodations to ListTile for scaled text.

gspencergoog opened the pull request:
This makes ListTile expand vertically when text is scaled, or really when whatever is placed inside it is larger than the available space.

Also, updated some doc comments to wrap at 80 cols, and changed one enum test to use a switch instead.

I had to modify the ListTile test to be more representative of the intention of the spec: we were testing with text in the leading and trailing sections, and the design wants icons there.  Because there was leading text, and the dense mode only changes the font size on the text lines, the leading text was propping up the minimum size of the tile.

Addresses #12311

gspencergoog committed:
Add accomodations to ListTile for scaled text.

gspencergoog committed:
Add accomodations to ListTile for scaled text.

gspencergoog committed:
Add scaled test.

gspencergoog committed:
Add scaled test.

Hixie commented on packages/flutter/lib/src/material/list_tile.dart:
```diff
@@ -351,9 +357,18 @@ class ListTile extends StatelessWidget {
+          break;
+      }
+    }
+    style ??= theme.textTheme.subhead;
```
this can just be an else clause, that way you avoid another branch. :-)

gspencergoog replied:
Done.

Hixie commented on packages/flutter/lib/src/material/list_tile.dart:
```diff
@@ -441,11 +456,13 @@ class ListTile extends StatelessWidget {
+        constraints: new BoxConstraints(minHeight: tileHeight),
+        child: new Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 16.0),
+          child: new IntrinsicHeight(child: new Row(children: children)),
```
We really REALLY want to avoid using IntrinsicHeight in the framework itself if at all possible, it's massively expensive.

Hixie replied:
(in the worst case it turns layout from linear to quadratic)

Why do you need an IntrinsicHeight here? Won't the Row just size itself to fit its children anyway?

gspencergoog replied:
You would think so.  But in list_tile_test.dart, it sizes itself to [800x600 here](https://github.com/flutter/flutter/blob/4e48a737eb4e0e689e52678a50d5f15ee26c5b0f/packages/flutter/test/material/list_tile_test.dart#L104), rather than 800x56, without the IntrinsicHeight.

This is another case of possibly breaking customer code.  I think that's probably more justified here than in the button widget, since it's less likely to simply contain a single widget, but I was still loathe to do that.

Hixie commented on packages/flutter/test/material/list_tile_test.dart:
```diff
@@ -45,57 +45,65 @@ void main() {
     bool hasSubtitle;
 
-    Widget buildFrame({ bool dense: false, bool isTwoLine: false, bool isThreeLine: false }) {
+    Widget buildFrame({bool dense: false, bool isTwoLine: false, bool isThreeLine: false, double textScaleFactor: 1.0}) {
```
flutter style has the spaces around named arguments

gspencergoog replied:
Done.

gspencergoog committed:
Review changes.

gspencergoog committed:
Review changes.

gspencergoog committed:
Fixing row height expansion

gspencergoog committed:
Fixing row height expansion

gspencergoog commented:
OK, I think I fixed the Row height expansion by wrapping it in a Column that has minimum expansion set.  This seems to pass the tests, and doesn't use IntrinsicHeight.

gspencergoog committed:
Add an UnconstrainedBox.

UnconstrainedBox will allow its child to size itself as if it had no constraints, and then attempt to fit around that object, until its own constraints are exceeded, in which case it will clip and display an overflow warning.

I also factored out DebugOverflowIndicator, which will draw overflow indicators on containers which overflow but aren't expected to.

gspencergoog committed:
Add an UnconstrainedBox.

UnconstrainedBox will allow its child to size itself as if it had no constraints, and then attempt to fit around that object, until its own constraints are exceeded, in which case it will clip and display an overflow warning.

I also factored out DebugOverflowIndicator, which will draw overflow indicators on containers which overflow but aren't expected to.

gspencergoog committed:
Add an UnconstrainedBox.

UnconstrainedBox will allow its child to size itself as if it had no constraints, and then attempt to fit around that object, until its own constraints are exceeded, in which case it will clip and display an overflow warning.

I also factored out DebugOverflowIndicator, which will draw overflow indicators on containers which overflow but aren't expected to.

Hixie commented on packages/flutter/lib/src/material/list_tile.dart:
```diff
@@ -441,11 +457,16 @@ class ListTile extends StatelessWidget {
+        constraints: new BoxConstraints(minHeight: tileHeight),
+        child: new Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 16.0),
+          child: new Column(
```
I don't understand why this Column is needed.

gspencergoog replied:
Because without it, the row expands vertically to fill the entire available space now that it is unconstrained in height.

This one is what happens with the Column (for a ListTile inside of a Center):
![screenshot_20171013-145340](https://user-images.githubusercontent.com/8867023/31568020-1f6050f6-b027-11e7-8a06-aff16f774baf.png)

And this is what happens without the Column (also a ListTile inside of a Center):
![screenshot_20171013-145403](https://user-images.githubusercontent.com/8867023/31568019-1f477ba8-b027-11e7-9093-5e3b0438c314.png)

If you never expect ListTiles to be used outside of a shrink-wrapped container, then the Column might not be needed.

Hixie replied:
We should find a better solution. Having a Column for a single child is not the optimal solution.

Hixie replied:
(Presumably that's UnconstrainedBox?)

Hixie commented:
Oh it's because the Column gives the child infinite max height, but the Center (even with heightFactor:1) gives a finite max height.

This is where that widget we'd discussed would come in useful (the "UnconstrainedBox" child that's sort of a hybrid between ConstrainedBox, Align, and OverflowBox).

gspencergoog commented:
Yes, it would.  I started to code that up a few days ago, but realized that I'm really not sure how it should work, and whether it was orthogonal to what already exists, or if it simply should be an additional parameter to an existing widget.  So I decided to punt to the Column for now so that we can solve the problem, and make it more efficient later.  Would it be OK to proceed with this PR as-is and code up the UnconstrainedBox later?

From our discussion before, I understand that UnconstrainedBox should:
- When the child is non-infinite in extent in a dimension, remove constraints so that it uses it's natural size.  What should it do if it is infinite? Just clip?
- If the child overflows, show errors like Flex does.
- Provide arguments to say how to align the contents, centering by default.

I wasn't sure if it should be able to only address a single axis (probably yes, but default to both), or if something else (OverflowBox?) already does most of this, and just needs an alignment argument.

I also wasn't sure how the overflow region is determined if the parent uses the size of the child to size itself: is it just the parent's constraints applied to the child's natural size region?

gspencergoog committed:
Adding tests.

gspencergoog committed:
Adding tests.

gspencergoog committed:
Review changes.

gspencergoog committed:
Review changes.

gspencergoog committed:
More review changes. Fixed a bunch of 'See also:'s

gspencergoog committed:
More review changes. Fixed a bunch of 'See also:'s

gspencergoog committed:
More review changes. Fixed a bunch of 'See also:'s

gspencergoog committed:
Extending RenderAligningShiftedBox instead, fixing example code.

gspencergoog committed:
Extending RenderAligningShiftedBox instead, fixing example code.

gspencergoog committed:
Review Changes.

gspencergoog committed:
Review Changes.

gspencergoog committed:
Review Changes.

gspencergoog committed:
Review Changes.

gspencergoog committed:
Review Changes.

gspencergoog committed:
Adding constrainedAxis to UnconstrainedBox

gspencergoog committed:
Adding constrainedAxis to UnconstrainedBox

gspencergoog committed:
Merge branch 'scale_list_tile' of https://github.com/gspencergoog/flutter into scale_list_tile

gspencergoog commented:
FYI, this won't really be ready for review until https://github.com/flutter/flutter/pull/12856 lands.
